features/256-auth-gear-interaction-overview-in-next/api.md

```diff
@@ -0,0 +1,126 @@
+* Keep client SDK simple, less interaction with auth gear directly.
+* More cloud functions, a developer should create corresponding cloud funtion to fulfill application logic.
+
+## Major changes
+
+* Remove admin API in client SDK.
+* Remove user profile support in auth gear.
+
+## Architecture overview
+
+![](https://i.imgur.com/GQbup7u.png)
+
+- Cloud function has its own `user` table and it should save user profile (gender, age, ..., anything other than auth infomation) here.
+- Auth gear's `user` table saves auth related information only.
+
+## Use cases
+
+### client SDK
+
+* should support to access current user's attributes after login/signup successfully.
+* should support change password
+* should support forgot password flow
+* should support verify user
+
+### cloud function
+
+* should support to indicate it is restricted to be invoked by authenticated user only.
+* should support to have a table with user_id as foreign key.
+* should support to access cloud funtion user's attribute in context.
+* should support to set a user's role, disabled state.
+* should support to query user.
+
+## Auth related information
+
+Auth gear would focus on auth related information only, they are
+
+```go
+AuthData        map[string]string
+SignUpAt        string
+LastSeenAt      string
+LastLoginAt     string
+Disabled        bool
+DisabledMessage string
+DisabledExpiry  string
+Roles           []string
+Verified        bool
+```
+
+## How to update user attributes in client SDK
+
+Since client SDK _**won't**_ have API support to update user attributes directly, developer should create their own cloud function for client SDK to update user attributes accordingly.
+
+![](https://i.imgur.com/Awrge8l.png)
+
+And followings APIs will be dropped in client SDK:
+
+- `adminDisableUser`
+- `adminenableUser`
+- `adminResetPassword`
+- `assignUserRole`
+- `fetchUserRole`
+- `revokeUserRole`
+- `setAdminRole`
+- `setDefaultRole`
+
+## How to update user attributes in clound function
+
+For restricted attributes, cloud function should use corresponding interface in auth gear to update.
+
+| Attribute | Interface |
+| -------- | -------- |
+| role | `POST /auth/role/admin`<br/>`POST /auth/role/assign`<br/>`POST /auth/role/default`<br/>`POST /auth/role/revoke` |
+| disable | `POST /auth/disable/set` |
+
+## How to indicated cloud function supports for authenticated user only
+
+Clound function should have to way to indicate it will be executed by authenticated user only. Following is an possible example:
+
+```javascript=
+const skygear = require('skygear');
+
+function welcome(req, res) {
+    res.end("welcome " + req.context.user.email);
+}
+
+module.exports = skygear.auth.authed(welcome);
+```
+
+## How support to user referenced table in cloud function
+
+Auth gear should have auth related hooks to cloud function, so cloud function should have enough information to create user referenced table. Following is a possible example:
```

Comment:
This one need to think carefully...

I think there are two types of hooks:

1. Hooks that's enough for users to implement their own auth DB, basically like [the one at auth0](https://auth0.com/docs/connections/database/custom-db/create-db-connection#step-2-create-database-action-scripts)
2. ONE simple hooks that just call and pass all attributes of the user whenever it is updated (basically like the one-way sync concept, except we don't write it to DB, but pass it as a JSON object to a Cloud Functions, something like that)